src/compiler/moduleSpecifiers.ts

```diff
@@ -174,13 +180,34 @@ namespace ts.moduleSpecifiers {
         const bundledPkgReference = bundledPackageName ? combinePaths(bundledPackageName, relativeToBaseUrl) : relativeToBaseUrl;
         const importRelativeToBaseUrl = removeExtensionAndIndexPostFix(bundledPkgReference, ending, compilerOptions);
         const fromPaths = paths && tryGetModuleNameFromPaths(removeFileExtension(bundledPkgReference), importRelativeToBaseUrl, paths);
-        const nonRelative = fromPaths === undefined ? importRelativeToBaseUrl : fromPaths;
+        const nonRelative = fromPaths === undefined ? importRelativeToBaseUrl : fromPaths.path;
 
         if (relativePreference === RelativePreference.NonRelative) {
             return nonRelative;
         }
 
-        if (relativePreference !== RelativePreference.Auto) Debug.assertNever(relativePreference);
+        if (relativePreference === RelativePreference.ExternalNonRelative) {
+            Debug.assertIsDefined(host.getNearestAncestorDirectoryWithPackageJson);

```

Comment:
Is this something that's true by construction, or are we checking for a misconfiguration?

Reply:
It was necessarily true when TS Server used it, and would never take this branch when the checker/emitter used it, but the assertion could have been violated by an API consumer, so I removed it and provided a fallback.